src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -15,66 +15,99 @@ public partial class Regex
     {
         public static int CacheSize
         {
-            get => RegexCache.CacheSize;
-            set => RegexCache.CacheSize = value;
+            get => RegexCache.MaxCacheSize;
+            set
+            {
+                if (value < 0)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value));
+                }
+
+                RegexCache.MaxCacheSize = value;
+            }
         }
     }
 
+    /// <summary>Cache used to store Regex instances used by the static methods on Regex.</summary>
     internal sealed class RegexCache
     {
+        // The implementation is optimized to make cache hits fast and lock-free, only taking a global lock
+        // when adding a new Regex to the cache.  Previous implementations of the cache took a global lock
+        // on all accesses, negatively impacting scalability, in order to minimize costs when the cache
+        // limit was hit and items needed to be dropped.  In such situations, however, we're having to
+        // pay the relatively hefty cost of creating a new Regex, anyway, and if the consuming app cares
+        // about such costs, it should either increase Regex.CacheSize or do its own Regex instance caching.
+
+        /// <summary>The default maximum number of items to store in the cache.</summary>
         private const int DefaultMaxCacheSize = 15;
-        private const int CacheDictionarySwitchLimit = 10;
-
-        private static readonly Dictionary<Key, Node> s_cache = new Dictionary<Key, Node>(DefaultMaxCacheSize);
-        private static Node? s_cacheFirst, s_cacheLast; // linked list for LRU and for small cache
+        /// <summary>The maximum number of cached items to examine when we need to replace an existing one in the cache with a new one.</summary>
+        /// <remarks>This is a somewhat arbitrary value, chosen to be small but at least as large as DefaultMaxCacheSize.</remarks>
+        private const int MaxExamineOnDrop = 30;
+
+        /// <summary>A read-through cache of one element, representing the most recently used regular expression.</summary>
+        private static volatile Node? s_lastAccessed;
+        /// <summary>The thread-safe dictionary storing all the items in the cache.</summary>
+        /// <remarks>
+        /// The concurrency level is initialized to 1 as we're using our own global lock for all mutations, so we don't need ConcurrentDictionary's
+        /// striped locking.  Capacity is initialized to 31, which is the same as (the private) ConcurrentDictionary.DefaultCapacity.
+        /// </remarks>
+        private static readonly ConcurrentDictionary<Key, Node> s_cacheDictionary = new ConcurrentDictionary<Key, Node>(concurrencyLevel: 1, capacity: 31);
+        /// <summary>A list of all the items in the cache.  Protected by <see cref="SyncObj"/>.</summary>
+        private static readonly List<Node> s_cacheList = new List<Node>(DefaultMaxCacheSize);
```

Comment:
What do we use the `s_cacheList` for? Just for knowing which Regexs to remove when `MaxCacheSize` is updated? Couldn't we do that be enumerating through the dictionary?

Reply:
> Just for knowing which Regexs to remove when MaxCacheSize is updated?

Yes

> Couldn't we do that by enumerating through the dictionary?

Yes.  I was very close to submitting the PR with that.  But then I looked at perf tests and backed off from it.  The problem is when the cache size is large.  If the cache size is the default size of 15, just enumerating the dictionary works quite well (enumerating a ConcurrentDictionary does require allocating an enumerator object, but if you're here you just allocated/parsed/maybe compiled/etc. a Regex, so an extra enumerator object is a drop in the bucket).  But if someone sets the cache size large, then our O(N) enumeration to compare every element ends up becoming potentially very expensive, especially in the degenerate case where you're using different regexs every time, and so effectively it becomes O(N^2).  This showed up as an ~6x perf hit in some of our perf tests.  So, I mitigated it by setting a limit on the number of elements we explore.  If it's under that limit (and the default cache size is), then we explore everything, but if it's over that limit, we explore only that number of elements.  We could just enumerate the dictionary, but we'll end up always looking at the first N items, which isn't going to be particularly well distributed over time.  So, I instead maintain this list that supports O(1) random access, and when we need to only look at a subset of its items, we look at a random subset distributed across the list.  I think this is reasonable for now; if we can come up with a better scheme, I'd be happy for us to switch in the future: my main concern is ensuring that cache reads remain lock-free and fast.